CardanoSharp.Wallet/CIPs/CIP2/CoinSelectionService.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using CardanoSharp.Wallet.CIPs.CIP2.ChangeCreationStrategies;
+using CardanoSharp.Wallet.Extensions;
+using CardanoSharp.Wallet.Extensions.Models.Transactions;
+using CardanoSharp.Wallet.Models;
+using CardanoSharp.Wallet.Models.Transactions;
+
+namespace CardanoSharp.Wallet.CIPs.CIP2
+{
+    public class CoinSelectionService
+    {
+        private readonly ICoinSelectionStrategy _coinSelection;
+        private readonly IChangeCreationStrategy _changeCreation;
+
+        public CoinSelectionService(ICoinSelectionStrategy coinSelection, IChangeCreationStrategy changeCreation)
+        {
+            _coinSelection = coinSelection;
+            _changeCreation = changeCreation;
+        }
+
+        public CoinSelection GetCoinSelection(List<TransactionOutput> outputs, List<Utxo> utxos, int limit = 20)
```

Comment:
Assuming this is the main entry point for Coin selection that our consumers will be using, it would be friendlier for them to use a different type other than the CDDL type TransactionOutput (with byte arrays)? Most upper layers get data from Blockfrost and Koios, they already work with friendlier types like hexadecimal or bech32 strings. The new Utxo type is nice though.

Also to give consumers more flexibility with parameters since they might already have a collection type different from `List<T>`, e.g. Arrays, IEnumerables, HashSets, ImmutableArrays etc. We can open up the parameter collection types to `IEnumerable<T>` or `IReadOnlyCollection<T>`? It will also give them a bit of assurance we won't mutate their collection.

Reply:
> it would be friendlier for them to use a different type other than the CDDL type TransactionOutput
so the way i saw this operation being used is during the Transaction Building phase. 
1. you added your outputs to your Tx Body with what you are trying to do 
2. you pulled UTxOs from Koios/Blockfrost
3. you call CoinSelection with your Outputs you already added and the utxos you got from your online piece.

i'm ok with making it friendlier depending a different workflow. I ultimately dont want to map to "friendlier" objects just to map it back to what it was originally.

>Also to give consumers more flexibility with parameters since they might already have a collection type different from List<T>, e.g. Arrays, IEnumerables, HashSets, ImmutableArrays etc. We can open up the parameter collection types to IEnumerable<T> or IReadOnlyCollection<T>? It will also give them a bit of assurance we won't mutate their collection.

good call. will make this change

Reply:
> > it would be friendlier for them to use a different type other than the CDDL type TransactionOutput
> 
> so the way i saw this operation being used is during the Transaction Building phase.
> 
> 1. you added your outputs to your Tx Body with what you are trying to do
> 2. you pulled UTxOs from Koios/Blockfrost
> 3. you call CoinSelection with your Outputs you already added and the utxos you got from your online piece.
> 
> i'm ok with making it friendlier depending a different workflow. I ultimately dont want to map to "friendlier" objects just to map it back to what it was originally.

I now see your point based on that sequence where the transaction body building happens first. My usage of CardanoSharp almost always has step 2 before 1, where I already have _friendly_ types and values (e.g. string addresses and policyIds) from Koios, Blockfrost or my own data store. Based on these values I then build the transaction to spend/mint/etc. so the CDDL-mapped types don't appear until later.

Either way there is no harm in catering for different use-cases as long as we can wok towards https://github.com/CardanoSharp/cardanosharp-wallet/pull/72#issuecomment-1157365465